File: eficiencia_algoritmica.cpp
//El programa genera distintos conjuntos de datos (aleatorios, casi ordenados, inversamente ordenados y con muchos duplicados) 
//en tres tamaños (100, 300 y 500 elementos) y los ordena utilizando un algoritmo QuickSort optimizado con la técnica de "mediana de tres" 
//y std::sort para subarrays pequeños. Para medir la eficiencia, el tiempo de ejecución se evalúa varias veces para cada conjunto de datos 
//y se calcula un promedio, proporcionando una medida precisa de la eficiencia del algoritmo en diferentes escenarios.
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>
#include <numeric> 

// Función para calcular la mediana de tres elementos y colocarla en la posición correcta
int medianaDeTres(std::vector<int>& arreglo, int izquierda, int derecha) {
    int centro = (izquierda + derecha) / 2;

    // Ordenar los tres elementos (izquierda, centro, derecha)
    if (arreglo[izquierda] > arreglo[centro]) std::swap(arreglo[izquierda], arreglo[centro]);
    if (arreglo[izquierda] > arreglo[derecha]) std::swap(arreglo[izquierda], arreglo[derecha]);
    if (arreglo[centro] > arreglo[derecha]) std::swap(arreglo[centro], arreglo[derecha]);

    // Colocar la mediana (centro) en la penúltima posición (derecha - 1) para usarla como pivote
    std::swap(arreglo[centro], arreglo[derecha - 1]);
    return arreglo[derecha - 1]; // Devolver el valor del pivote
}

// Función principal de QuickSort (recursiva)
void ordenarQuicksort(std::vector<int>& arreglo, int izquierda, int derecha) {
    // Caso base: si el subarreglo tiene menos de 10 elementos, usar std::sort (más eficiente para subarreglos pequeños)
    if (izquierda + 10 <= derecha) {
        int pivote = medianaDeTres(arreglo, izquierda, derecha); // Seleccionar el pivote

        // Particionar el arreglo alrededor del pivote
        int indiceIzquierda = izquierda;
        int indiceDerecha = derecha - 1;
        while (true) {
            while (arreglo[++indiceIzquierda] < pivote) {} // Mover indiceIzquierda hasta encontrar un elemento mayor o igual al pivote
            while (arreglo[--indiceDerecha] > pivote) {}  // Mover indiceDerecha hasta encontrar un elemento menor o igual al pivote
            if (indiceIzquierda < indiceDerecha) {
                std::swap(arreglo[indiceIzquierda], arreglo[indiceDerecha]); // Intercambiar elementos si están en el lado incorrecto del pivote
            } else {
                break; // Salir del bucle cuando los índices se cruzan
            }
        }
        std::swap(arreglo[indiceIzquierda], arreglo[derecha - 1]); // Colocar el pivote en su posición final

        // Ordenar recursivamente las dos particiones
        ordenarQuicksort(arreglo, izquierda, indiceIzquierda - 1);
        ordenarQuicksort(arreglo, indiceIzquierda + 1, derecha);
    } else {
        std::sort(arreglo.begin() + izquierda, arreglo.begin() + derecha + 1); // Ordenar subarreglo pequeño con std::sort
    }
}

// Función para medir el tiempo de ejecución promedio de QuickSort sobre varias repeticiones
void medirTiempoDeOrdenamiento(std::vector<int>& arreglo, int repeticiones) {
    std::vector<double> tiempos(repeticiones); // Vector para almacenar los tiempos de cada repetición

    for (int i = 0; i < repeticiones; ++i) {
        std::vector<int> copiaArreglo = arreglo; // Crear una copia del arreglo para no modificar el original
        auto inicio = std::chrono::high_resolution_clock::now(); // Registrar el tiempo de inicio
        ordenarQuicksort(copiaArreglo, 0, copiaArreglo.size() - 1); // Ordenar la copia
        auto fin = std::chrono::high_resolution_clock::now(); // Registrar el tiempo de fin
        std::chrono::duration<double> duracion = fin - inicio; // Calcular la duración
        tiempos[i] = duracion.count(); // Almacenar el tiempo en segundos
    }

    // Calcular el tiempo promedio
    double tiempoPromedio = std::accumulate(tiempos.begin(), tiempos.end(), 0.0) / repeticiones;
    std::cout << "Tiempo de ejecución promedio: " << tiempoPromedio << " segundos\n";
}

// Función para generar datos aleatorios
std::vector<int> generarDatosAleatorios(int tamano, int rango) {
    std::vector<int> datos(tamano);
    std::generate(datos.begin(), datos.end(), [rango]() { return rand() % rango; }); // Generar números aleatorios en el rango [0, rango)
    return datos;
}

// Función para generar datos casi ordenados
std::vector<int> generarDatosCasiOrdenados(int tamano) {
    std::vector<int> datos(tamano);
    std::iota(datos.begin(), datos.end(), 0); // Generar una secuencia ordenada de 0 a tamano-1
    // Intercambiar algunos elementos para desordenar ligeramente
    for (int i = 0; i < tamano / 10; ++i) {
        std::swap(datos[rand() % tamano], datos[rand() % tamano]); 
    }
    return datos;
}

// Función para generar datos inversamente ordenados
std::vector<int> generarDatosInversamenteOrdenados(int tamano) {
    std::vector<int> datos(tamano);
    std::iota(datos.rbegin(), datos.rend(), 0); // Generar una secuencia inversa de tamano-1 a 0
    return datos;
}

// Función para generar datos con muchos duplicados
std::vector<int> generarDatosConDuplicados(int tamano, int rango) {
    std::vector<int> datos(tamano);
    std::generate(datos.begin(), datos.end(), [rango]() { return rand() % (rango / 10); }); // Generar números aleatorios con muchos duplicados
    return datos;
}

int main() {
    int tamanos[] = {100, 300, 500}; // Tamaños de los conjuntos de datos
    int repeticiones = 5; // Número de repeticiones para cada conjunto de datos

    for (int tamano : tamanos) {
        std::cout << "Conjunto de datos aleatorios (" << tamano << " elementos):\n";
        std::vector<int> datosAleatorios = generarDatosAleatorios(tamano, 1000);
        medirTiempoDeOrdenamiento(datosAleatorios, repeticiones);

        std::cout << "Conjunto de datos casi ordenados (" << tamano << " elementos):\n";
        std::vector<int> datosCasiOrdenados = generarDatosCasiOrdenados(tamano);
        medirTiempoDeOrdenamiento(datosCasiOrdenados, repeticiones);

        std::cout << "Conjunto de datos inversamente ordenados (" << tamano << " elementos):\n";
        std::vector<int> datosInversamenteOrdenados = generarDatosInversamenteOrdenados(tamano);
        medirTiempoDeOrdenamiento(datosInversamenteOrdenados, repeticiones);

        std::cout << "Conjunto de datos con muchos duplicados (" << tamano << " elementos):\n";
        std::vector<int> datosConDuplicados = generarDatosConDuplicados(tamano, 1000);
        medirTiempoDeOrdenamiento(datosConDuplicados, repeticiones);

        std::cout << "-----------------------------------\n";
    }

    return 0;
}

File: eficiencia_algoritmica.py
import random
import time
import psutil
import os

def mediana_de_tres(arreglo, bajo, alto):

    #Selecciona el pivote usando la técnica de la mediana de tres.
  
    medio = (bajo + alto) // 2
    if arreglo[bajo] > arreglo[medio]:
        arreglo[bajo], arreglo[medio] = arreglo[medio], arreglo[bajo]
    if arreglo[bajo] > arreglo[alto]:
        arreglo[bajo], arreglo[alto] = arreglo[alto], arreglo[bajo]
    if arreglo[medio] > arreglo[alto]:
        arreglo[medio], arreglo[alto] = arreglo[alto], arreglo[medio]
    return medio

def particionar(arreglo, bajo, alto):
    """
    Función para particionar el arreglo en el algoritmo QuickSort.
    Utiliza la mediana de tres para seleccionar el pivote.
    """
    pivote_indice = mediana_de_tres(arreglo, bajo, alto)
    pivote = arreglo[pivote_indice]
    arreglo[pivote_indice], arreglo[alto] = arreglo[alto], arreglo[pivote_indice]
    
    i = bajo - 1
    for j in range(bajo, alto):
        if arreglo[j] <= pivote:
            i += 1
            arreglo[i], arreglo[j] = arreglo[j], arreglo[i]
    arreglo[i + 1], arreglo[alto] = arreglo[alto], arreglo[i + 1]
    return i + 1

def insercion_directa(arreglo, bajo, alto):

    #Implementación del algoritmo de ordenamiento por inserción directa.

    for i in range(bajo + 1, alto + 1):
        clave = arreglo[i]
        j = i - 1
        while j >= bajo and arreglo[j] > clave:
            arreglo[j + 1] = arreglo[j]
            j -= 1
        arreglo[j + 1] = clave

def ordenar_quicksort(arreglo, bajo, alto):
    """
    Implementación optimizada del algoritmo QuickSort.
    Utiliza inserción directa para subarreglos pequeños y evita la recursión excesiva.
    """
    while bajo < alto:
        if alto - bajo + 1 < 10:
            insercion_directa(arreglo, bajo, alto)
            break
        else:
            indice_particion = particionar(arreglo, bajo, alto)
            if indice_particion - bajo < alto - indice_particion:
                ordenar_quicksort(arreglo, bajo, indice_particion - 1)
                bajo = indice_particion + 1
            else:
                ordenar_quicksort(arreglo, indice_particion + 1, alto)
                alto = indice_particion - 1

def obtener_uso_memoria():

    #Función para obtener el uso actual de memoria en MB.
 
    proceso = psutil.Process(os.getpid())
    return proceso.memory_info().rss / (1024 * 1024)  # Convertir a MB

def medir_rendimiento(arreglo):

    #Mide el tiempo de ejecución y uso de memoria del algoritmo QuickSort.
  
    copia_arreglo = arreglo.copy()
    
    tiempo_inicio = time.time()
    memoria_inicio = obtener_uso_memoria()
    
    ordenar_quicksort(copia_arreglo, 0, len(copia_arreglo) - 1)
    
    tiempo_fin = time.time()
    memoria_fin = obtener_uso_memoria()
    
    tiempo_ejecucion = tiempo_fin - tiempo_inicio
    uso_memoria = memoria_fin - memoria_inicio
    
    return tiempo_ejecucion, uso_memoria, copia_arreglo

def generar_arreglo_aleatorio(tamano):

    #Genera un arreglo de números enteros aleatorios.

    return [random.randint(1, 1000) for _ in range(tamano)]

def verificar_ordenamiento(arreglo):

    #Verifica si el arreglo está correctamente ordenado.

    return all(arreglo[i] <= arreglo[i+1] for i in range(len(arreglo)-1))

def main():
    tamanos = [100, 300, 500]
    
    for tamano in tamanos:
        print(f"\nConjunto de {tamano} elementos:")
        arreglo_original = generar_arreglo_aleatorio(tamano)
        
        tiempo_ejecucion, uso_memoria, arreglo_ordenado = medir_rendimiento(arreglo_original)
        
        print(f"Tiempo de Ejecución: {tiempo_ejecucion:.6f} segundos")
        print(f"Uso de Memoria: {uso_memoria:.6f} MB")
        
        if verificar_ordenamiento(arreglo_ordenado):
            print("El arreglo está correctamente ordenado.")
        else:
            print("Error: El arreglo no está correctamente ordenado.")

if __name__ == "__main__":
    main()